src/Data/Text.hs

```diff
@@ -1623,8 +1533,11 @@ group = groupBy (==)
 -- | /O(n)/ Return all initial segments of the given 'Text', shortest
 -- first.
 inits :: Text -> [Text]
-inits t@(Text arr off len) = loop 0
-    where loop i | i >= len = [t]
+inits t = empty : if null t then [] else go t
+    where 
+      go (Text arr off len) = loop (iter_ t 0)
+        where
+          loop i | i >= len = [t]
```

Comment:
It can be condensed further:

```
inits :: Text -> [Text]
inits t = empty : loop 0
  where
    Text arr off len = t
    loop i | i >= len = []
           | otherwise = let j = i + iter_ t i in Text arr off j : loop j
```

If the core doesn't look good, you can also try `empty : case t of ...` and define the `loop` in a `let`.

Reply:
Neat! Here's the core that gives us. I think it looks alright. It's turning `arr` and `off` into selector thunks but that should be fine imo

<section>
<summary> Core </summary>

```
 
23063   │ inits
23064   │   = \ t_a69c ->
23065   │       : empty
23066   │         (let {
23067   │            arr_sbsA
23068   │              = case t_a69c of { Text bx_dafw bx1_dafx bx2_dafy ->
23069   │                ByteArray bx_dafw
23070   │                } } in
23071   │          let {
23072   │            off_sbsz
23071   │          let {
23072   │            off_sbsz
23073   │              = case t_a69c of { Text bx_dafz bx1_dafA bx2_dafB ->
23074   │                I# bx1_dafA
23075   │                } } in
23076   │          case t_a69c of { Text bx_dafC bx1_dafD bx2_dafE ->
23077   │          case >=# 0# bx2_dafE of {
23078   │            __DEFAULT ->
23079   │              let {
23080   │                j_sbsD
23081   │                  = case indexWord8Array# bx_dafC bx1_dafD of r#_ibhd { __DEFAULT ->
23082   │                    let {
23083   │                      c#_sbsC
23084   │                        = word2Int# (clz8# (and# (not# (word8ToWord# r#_ibhd)) 255##)) } in
23085   │                    I# (xorI# c#_sbsC (<=# c#_sbsC 0#))
23086   │                    } } in
23087   │              : ($WText arr_sbsA off_sbsz j_sbsD)
23088   │                (case j_sbsD of { I# ww_X3 ->
23089   │                 letrec {
23090   │                   $wloop_sdCw
23091   │                     = \ ww1_sdCt ->
23092   │                         case >=# ww1_sdCt bx2_dafE of {
23093   │                           __DEFAULT ->
23094   │                             let {
23095   │                               j1_X5
23096   │                                 = case indexWord8Array# bx_dafC (+# bx1_dafD ww1_sdCt) of r#_ibhd
23097   │                                   { __DEFAULT ->
23098   │                                   let {
23099   │                                     c#_sbsC
23100   │                                       = word2Int#
23101   │                                           (clz8# (and# (not# (word8ToWord# r#_ibhd)) 255##)) } in
23102   │                                   I# (+# ww1_sdCt (xorI# c#_sbsC (<=# c#_sbsC 0#)))
23103   │                                   } } in
23104   │                             : ($WText arr_sbsA off_sbsz j1_X5)
23105   │                               (case j1_X5 of { I# ww2_X7 -> $wloop_sdCw ww2_X7 });
23106   │                           1# -> []
23107   │                         }; } in
23108   │                 $wloop_sdCw ww_X3
23109   │                 });
23110   │            1# -> []
23111   │          }
```

</section>

Reply:
I was hoping that they would be unboxed before entering the loop. So an explicit `case t of` would be better.

Reply:
New Core:
```
inits
19985   │   = \ t ->
19986   │       : empty
19987   │         (case t of { Text bx bx1 bx2 ->
19988   │          letrec {
19989   │            $wloop
19990   │              = \ ww ->
19991   │                  case >=# ww bx2 of {
19992   │                    __DEFAULT ->
19993   │                      let {
19994   │                        j = case indexWord8Array# bx (+# bx1 ww) of r# { __DEFAULT ->
19995   │                            let {
19996   │                              c# = word2Int# (clz8# (and# (not# (word8ToWord# r#)) 255##)) } in
19997   │                            I# (+# ww (xorI# c# (<=# c# 0#)))
19998   │                            } } in
19999   │                      : (case j of { I# unbx -> Text bx bx1 unbx })
20000   │                        (case j of { I# ww1 -> $wloop ww1 });
20001   │                    1# -> []
20002   │                  }; } in
20003   │          $wloop 0#
20004   │          })
```

I think I want to add a bang to `j`

Reply:
And Core with a bang on j:
```
inits
19985   │   = \ t ->
19986   │       : empty
19987   │         (case t of { Text bx bx1 bx2 ->
19988   │          letrec {
19989   │            $wloop
19990   │              = \ ww ->
19991   │                  case >=# ww bx2 of {
19992   │                    __DEFAULT ->
19993   │                      case indexWord8Array# bx (+# bx1 ww) of r# { __DEFAULT ->
19994   │                      let {
19995   │                        c# = word2Int# (clz8# (and# (not# (word8ToWord# r#)) 255##)) } in
19996   │                      let { j = +# ww (xorI# c# (<=# c# 0#)) } in
19997   │                      : (Text bx bx1 j) ($wloop j)
19998   │                      };
19999   │                    1# -> []
20000   │                  }; } in
20001   │          $wloop 0#
20002   │          })
```

Reply:
Good stuff! Looks good to me!